### PR TITLE
ft: DataDog adaptor

### PR DESCRIPTION
**Pull request by hauleth**
This introduces wrapper adaptor over Webhook adaptor to send logs to
DataDog HTTP log ingest endpoint.

### CODE REVIEW

File: test/logflare/backends/adaptor/datadog_adaptor_test.exs
Comment (Ziinc) [inline test/logflare/backends/adaptor/datadog_adaptor_test.exs:113]: tags need to be a certain format
https://docs.datadoghq.com/getting_started/tagging/#define-tags
would need to handle restrictions for them.

File: lib/logflare/backends/adaptor/datadog_adaptor.ex
Comment (Ziinc) [inline lib/logflare/backends/adaptor/datadog_adaptor.ex:97]: retry with backoff would be great

Comment (Ziinc) [reply lib/logflare/backends/adaptor/datadog_adaptor.ex:97]: we can add it after we add in batching for webhook adaptor

File: lib/logflare/backends/adaptor/datadog_adaptor.ex
Comment (Ziinc) [inline lib/logflare/backends/adaptor/datadog_adaptor.ex:30]: This would get called for even small requests or singular event requests, as `Backends.ingest_logs/2` doesn't handle batching.
would need to handle batching at adaptor level, which is why i've mostly been using Broadway to auto-handle them. This won't survive at prod scale, which is the issues facing the current WebhookAdaptor and PostgresAdaptor.

**Review by Ziinc — CHANGES_REQUESTED** [review]

File: test/logflare/backends/adaptor/datadog_adaptor_test.exs
Comment (hauleth) [reply test/logflare/backends/adaptor/datadog_adaptor_test.exs:113]: Ok, I will take a look. Maybe there should a way to define mappings for tags to avoid excessive tag amount as docs mention:

> Tags should not originate from unbounded sources, such as epoch timestamps, user IDs, or request IDs. Doing so may infinitely [increase the number of metrics](https://docs.datadoghq.com/metrics/) for your organization and impact your billing.

File: lib/logflare/backends/adaptor/datadog_adaptor.ex
Comment (hauleth) [reply lib/logflare/backends/adaptor/datadog_adaptor.ex:97]: Yeah. That module probably should be extracted to be independent of the DataDog adaptor, or even to separate library, but for now I wanted to have everything simple and in single place.

File: lib/logflare/backends/adaptor/datadog_adaptor.ex
Comment (hauleth) [reply lib/logflare/backends/adaptor/datadog_adaptor.ex:30]: Yesm that would get called for singular events, that is why that module *is* implementing batching in a way that it will wait until there is given amount of events in queue **or** timeout was reached. The goal of the timeout is to not batch messages indefinitely if there will be not enough of them, but rather send messages in more or less uniform periods even if there is less than N messages.

Comment (Ziinc) [reply lib/logflare/backends/adaptor/datadog_adaptor.ex:30]: this is the same as the `:batch_timeout` option for broadway

File: test/logflare/backends/adaptor/datadog_adaptor_test.exs
Comment (Ziinc) [reply test/logflare/backends/adaptor/datadog_adaptor_test.exs:113]: yea i'm thinking of implementing a allowlist for schema paths that are permitted for schema updates, it can take a similar form:
```
metadata\..+
my
field
```
was thinking of making it regex based, can do a similar allowlist for allowed tag mappings

File: lib/logflare/backends/adaptor/datadog_adaptor.ex
Comment (hauleth) [reply lib/logflare/backends/adaptor/datadog_adaptor.ex:30]: Oh, TIL. I can change it to Broadway to make everything more uniform

File: lib/logflare/backends/adaptor/datadog_adaptor.ex
Comment (Ziinc) [inline lib/logflare/backends/adaptor/datadog_adaptor.ex:29]: nit: move to module attribute with inline comments to datadog documentation

File: lib/logflare/backends/adaptor/datadog_adaptor.ex
Comment (Ziinc) [inline lib/logflare/backends/adaptor/datadog_adaptor.ex:61]:
```suggestion
    changeset
    |> Ecto.Changeset.validate_required([:api_key])
```

**Review by Ziinc — CHANGES_REQUESTED** [review]
Implementation looks good, small nits, good to merge after they're cleared.

For further improvements down the road, we'd need to consider their batch sizes:
Screenshot 2024-04-25 at 1 32 51 AM
we'd likely need configurable max batch size and max batch length for webhook adaptor, similar to what [BQ adaptor is doing](https://github.com/Logflare/logflare/commit/8e9d51a7e40501f42c248d5181e82e648e023adf)
will backlog this until their api starts rejecting us

**Review by Ziinc — DISMISSED** [review]

**Comment by hauleth** [comment]
@Ziinc what happened with all these commits?

**Comment by Ziinc** [comment]
Ah I rebased it in the log drain PR

**Comment by hauleth** [comment]
Should I clean this up with my local state?

--

Łukasz Niemier
***@***.***

**Comment by Ziinc** [comment]
Have pushed up the region changes

Will need to do some more testing before merge

- [ ] manual test for DD drain rule

**Review by Ziinc — APPROVED** [review]

**Comment by hauleth** [comment]
💚 💙 💜 💛 ❤️

**Pull request merged**
